From 0a086cc381baeea565d67447d09c6b03351667f5 Mon Sep 17 00:00:00 2001
From: Olivier Moysan <olivier.moysan@foss.st.com>
Date: Thu, 9 Oct 2025 09:21:28 +0200
Subject: [PATCH] counter: stm32-timer-cnt: add capture input selector

There is an input capture/compare selector on each timer channel, that
allows to choose between tixfp1, tixfp2 or trc signals to provide icx
signal. This muxer can be seen as a capture attribute.
Expose a "capture_source" attribute for each capture path to select the
source signal. The sources are identified by an index where 0 stands for
tixfpn and 1 for tiyfpn (n=[1..4]). For instance for channel 1, either
ti1fp1 or ti2fp1 can be selected.

example of use:
cd /sys/bus/counter/devices/counter0/count0
echo 1 > capture_source0

Change-Id: I518ee047da63acf093611393a5baacc746725961
Signed-off-by: Olivier Moysan <olivier.moysan@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/501440
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
---
 drivers/counter/stm32-timer-cnt.c | 122 +++++++++++++++++++++++++-----
 1 file changed, 104 insertions(+), 18 deletions(-)

--- a/drivers/counter/stm32-timer-cnt.c
+++ b/drivers/counter/stm32-timer-cnt.c
@@ -339,6 +339,108 @@ static int stm32_count_nb_ovf_write(stru
 	return 0;
 }
 
+struct stm32_count_cc_regs {
+	u32 ccmr_reg;
+	u32 ccmr_mask;
+	u32 ccmr_bits;
+	u32 ccer_bits;
+};
+
+static const struct stm32_count_cc_regs stm32_cc[] = {
+	{ TIM_CCMR1, TIM_CCMR_CC1S, TIM_CCMR_CC1S_TI1,
+		TIM_CCER_CC1E | TIM_CCER_CC1P | TIM_CCER_CC1NP },
+	{ TIM_CCMR1, TIM_CCMR_CC2S, TIM_CCMR_CC2S_TI2,
+		TIM_CCER_CC2E | TIM_CCER_CC2P | TIM_CCER_CC2NP },
+	{ TIM_CCMR2, TIM_CCMR_CC3S, TIM_CCMR_CC3S_TI3,
+		TIM_CCER_CC3E | TIM_CCER_CC3P | TIM_CCER_CC3NP },
+	{ TIM_CCMR2, TIM_CCMR_CC4S, TIM_CCMR_CC4S_TI4,
+		TIM_CCER_CC4E | TIM_CCER_CC4P | TIM_CCER_CC4NP },
+};
+
+static int stm32_capture_source_read(struct counter_device *counter,
+				     struct counter_count *count,
+				     size_t idx, u64 *source)
+{
+	struct stm32_timer_cnt *const priv = counter_priv(counter);
+	int ret;
+	u32 val;
+
+	ret = regmap_read(priv->regmap, stm32_cc[idx].ccmr_reg, &val);
+	if (ret)
+		return ret;
+
+	switch (idx) {
+	case 0:
+		*source = FIELD_GET(TIM_CCMR_CC1S, val);
+		break;
+	case 1:
+		*source = FIELD_GET(TIM_CCMR_CC2S, val);
+		break;
+	case 2:
+		*source = FIELD_GET(TIM_CCMR_CC3S, val);
+		break;
+	case 3:
+		*source = FIELD_GET(TIM_CCMR_CC4S, val);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/*
+	 * If *source is zero that means that the source is not yet configured.
+	 * Just return default source index in this case, as it will be used by default if not
+	 * explicitly configured.
+	 */
+	if (*source)
+		(*source)--;
+
+	if (*source > TIM_CCMR_CC1S_TI2 - 1)
+		return -EINVAL;
+
+	return 0;
+}
+
+static int stm32_capture_source_write(struct counter_device *counter,
+				      struct counter_count *count,
+				      size_t idx, u64 source)
+{
+	struct stm32_timer_cnt *const priv = counter_priv(counter);
+	u32 reg, mask, val;
+
+	if (source > TIM_CCMR_CC1S_TI2 - 1)
+		return -EINVAL;
+
+	switch (idx) {
+	case 0:
+		val = FIELD_PREP(TIM_CCMR_CC1S, source + 1);
+		reg = TIM_CCMR1;
+		mask = TIM_CCMR_CC1S;
+		break;
+	case 1:
+		val = FIELD_PREP(TIM_CCMR_CC2S, source + 1);
+		reg = TIM_CCMR1;
+		mask = TIM_CCMR_CC2S;
+		break;
+	case 2:
+		val = FIELD_PREP(TIM_CCMR_CC3S, source + 1);
+		reg = TIM_CCMR2;
+		mask = TIM_CCMR_CC3S;
+		break;
+	case 3:
+		val = FIELD_PREP(TIM_CCMR_CC4S, source + 1);
+		reg = TIM_CCMR2;
+		mask = TIM_CCMR_CC4S;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	regmap_update_bits(priv->regmap, reg, mask, val);
+
+	return 0;
+}
+
+static DEFINE_COUNTER_ARRAY_U64(stm32_count_capture_sources, 4);
 static DEFINE_COUNTER_ARRAY_CAPTURE(stm32_count_cap_array, 4);
 
 static struct counter_comp stm32_count_ext[] = {
@@ -349,6 +451,8 @@ static struct counter_comp stm32_count_e
 	COUNTER_COMP_COUNT_U64("prescaler", stm32_count_prescaler_read,
 			       stm32_count_prescaler_write),
 	COUNTER_COMP_ARRAY_CAPTURE(stm32_count_cap_read, NULL, stm32_count_cap_array),
+	COUNTER_COMP_COUNT_ARRAY_U64("capture_source", stm32_capture_source_read,
+				     stm32_capture_source_write, stm32_count_capture_sources),
 	COUNTER_COMP_COUNT_U64("num_overflows", stm32_count_nb_ovf_read, stm32_count_nb_ovf_write),
 };
 
@@ -407,24 +511,6 @@ static int stm32_action_read(struct coun
 	}
 }
 
-struct stm32_count_cc_regs {
-	u32 ccmr_reg;
-	u32 ccmr_mask;
-	u32 ccmr_bits;
-	u32 ccer_bits;
-};
-
-static const struct stm32_count_cc_regs stm32_cc[] = {
-	{ TIM_CCMR1, TIM_CCMR_CC1S, TIM_CCMR_CC1S_TI1,
-		TIM_CCER_CC1E | TIM_CCER_CC1P | TIM_CCER_CC1NP },
-	{ TIM_CCMR1, TIM_CCMR_CC2S, TIM_CCMR_CC2S_TI2,
-		TIM_CCER_CC2E | TIM_CCER_CC2P | TIM_CCER_CC2NP },
-	{ TIM_CCMR2, TIM_CCMR_CC3S, TIM_CCMR_CC3S_TI3,
-		TIM_CCER_CC3E | TIM_CCER_CC3P | TIM_CCER_CC3NP },
-	{ TIM_CCMR2, TIM_CCMR_CC4S, TIM_CCMR_CC4S_TI4,
-		TIM_CCER_CC4E | TIM_CCER_CC4P | TIM_CCER_CC4NP },
-};
-
 static int stm32_count_capture_configure(struct counter_device *counter, unsigned int ch,
 					 bool enable)
 {
